### ISSUE TITLE
How to cross compile for ARM with MSVC?

### ISSUE THREAD

**Issue by mstorsjo** [mesonbuild/meson#4402]
Normally when compiling with MSVC, you select the cross compilation target ("host" in the cross file) by running a script that sets up the environment (`PATH`, `LIB`) according to the desired host environment; the `cl.exe` in path is then the one for the chosen architecture, and `LIB` points to the directories containing libraries for this architecture.

If trying to run meson in such an environment, the default native compiler is also detected as `cl` from the path, a dialog box saying "The image file c:\...\sanitycheckc.exe is valid, but is for a machine type other than the current machine" appears, and meson fails with the following error message:

```
meson.build:25:0: ERROR:  Could not invoke sanity test executable: [WinError 216] This version of %1 is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher.
```

How is one supposed to set up cross compilation with Meson with the MSVC compiler, when you can only have one compiler in `PATH` and one `LIB` set of directories in the environment at a time?

**Comment by jon-turney** [comment]
I think you must use a [cross-file](http://mesonbuild.com/Cross-compilation.html) to specify the cross-toolchain.

**Comment by mstorsjo** [comment]
Ok, how do I specify a different `LIB` env var for the cross compiler vs the native one then?

**Comment by dcbaker** [comment]
It might be worth it to just create the awful wrappers ourselves and ship them with meson to avoid this coming up again and again?

**Comment by jon-turney** [comment]
Having a fix for #2392 would help, if we don't actually need to use the host compiler...

**Comment by mstorsjo** [comment]
> Yes, this is terrible on many different levels but that is sadly the current limitation we have.

Ok, thanks for spelling it out!

Btw, a minor addition to your blogpost about awful things, on the topic of spawning processes. In addition to the quoting mess, the exec() functions are highly misleading, as they look like their posix counterparts, but doesn't behave the same (as you can't replace the current process, exec() spawns a new child process, but doesn't wait for it to exit, and instead exits the current process with a zero return code).

> Having a fix for #2392 would help, if we don't actually need to use the host compiler...

Yup, for at least my current case, this would be enough.

**Comment by mstorsjo** [comment]
> * create wrapper bat files for cl and linker that first invoke the cross toolchain's `vcvarsall.bat` and then the actual cross compiler and put those in the cross file

How do I specify to use this wrapper linker bat file as linker for the cross compilation in the cross file? As far as I can see, `VisualStudioCCompiler` is hardcoded to just use `link`.

**Comment by jon-turney** [comment]
> How do I specify to use this wrapper linker bat file as linker for the cross compilation in the cross file? As far as I can see, `VisualStudioCCompiler` is hardcoded to just use `link`.

This is a defect :(

https://github.com/mesonbuild/meson/blob/4a9fd0e742e0cc9fb1607c182904ea48848ea1ba/mesonbuild/compilers/c.py#L1340-L1341

**Comment by mstorsjo** [comment]
Also, to elaborate on the FIXME comment: Picking `link` from the same directory as the compiler would be doable, but also quite cumbersome.

Consider if you have one directory that you add to the path, with wrapper bat files like `x86-cl.bat`, `arm64-cl.bat` etc - you can't have it just pick `link` from this directory. Making that work would require one separate directory for each architecture again, with e.g. `arm64-cl.bat` and `link.bat` in that one, `x86-cl.bat` and `link.bat` in another one, etc. And that directory would have to be moved to the end of `%PATH%` so that the default native one still is first.

Being able to override the linker name via the cross file would be much more straightforward.

**Comment by jon-turney** [comment]
Trying to write a test case for #4463/#4040, it seems that `lib` defaults `/machine` to the native architecture, so this problem is only going to show up when cross-compiling.

@mstorsjo I wonder if you could share your setup for cross-compiling to Windows ARM? I'd like to try to add something like that to our CI...

**Comment by jon-turney** [comment]
> * create wrapper bat files for cl and linker that first invoke the cross toolchain's `vcvarsall.bat` and then the actual cross compiler and put those in the cross file

It did occur to me that we could perhaps simplify this a bit by adding an `[env]` section to the cross/native file, which defines variables that are added to the environment when running anything from it's `[binaries]` section.  This could be used to set LIB, INCLUDE etc. for MSVC tools, thus at least avoiding the need for wrapper .bat files as well.

**Comment by jpakkane** [comment]
No. No environment variables! They are of the devil. In addition to being conceptually terrible they can only be added by spawning executables via a Python wrapper binary. This doubles the amount of processes to run, which is sloooooooow on Windows.

**Comment by Neumann-A** [comment]
please merge  #7190 so that we can actually compile for `arm(64)-windows` in vcpkg 

>On the other hand Meson's cross compilation support has been designed from the ground up to require that. The reason for this is that then you can compile code generators and run them transparently during the build in a single pass

That should be more an opt-in than an requirement. 
It should always be possible to have a two step process or supply the code generators in another way. (e.g. prebuilt). Why? If you have to build for a lot of different architectures it seems unnecessary to recompile the code generator every time. 
As such please fix mesons behavior ....

**Comment by jon-turney** [comment]
So, as of #7718, the headline issue here should be fixed, and we should have testing to cover it.

However, as mentioned above, the setup needed to successfully build something from a meson.build which uses both the native and cross-compiler using MS tools is complex.

> It might be worth it to just create the awful wrappers ourselves and ship them with meson to avoid this coming up again and again?

Now we have [constants] in machine files, perhaps something like the following would work? (untested):

crossfile
``` 
[constants]
envwrapper = ['envwrapper.bat', 'amd64_arm64']

[binaries]
c = envwrapper + ['cl']
etc.
```

nativefile
``` 
[constants]
envwrapper = ['envwrapper.bat', 'amd64']

[binaries]
c = envwrapper + ['cl']
etc.
```

where `envwrapper.bat` is a script which invokes `vcvars.bat` to set the specified compilation environment, then invokes the given command in that environment.

**Comment by bonzini** [comment]
Leaving this open as an RFE to have something for `vcvarsall.bat` that resembles `meson env2mfile`.
